### PR TITLE
KAFKA-14767: Fix missing commitId build error after git gc

### PR DESCRIPTION
**Pull request by gharris1727**
`git gc` moves commit hashes from individual `.git/refs/heads/` to `.git/packed-refs` which is not read by the determineCommitId function.

Replace the existing lookup within the `.git` directory with a GrGit lookup that handles packed and unpacked refs transparently.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -159,16 +159,9 @@ def determineCommitId() {
...
-    if (headRef.contains('ref: ')) {
-      headRef = headRef.replaceAll('ref: ', '').trim()
-      if (file("$rootDir/.git/$headRef").exists()) {
-        file("$rootDir/.git/$headRef").text.trim().take(takeFromHash)
-      }
-    } else {
-      headRef.trim().take(takeFromHash)
-    }
+  } else if (file("$rootDir/.git").exists()) {
+    def repo = Grgit.open(currentDir: project.getRootDir())
```
Comment (ijuma) [inline build.gradle]: The `rat` configuration does something similar, maybe we should have a variable for the git repo that can be used by both tasks.

**Comment by ijuma** [comment]
Looks like the build failed.

**Comment by gharris1727** [comment]
@ijuma I don't know how my change would impact the execution order of rat vs processTestMessages, do you have any intuition?

The commitId change that I originally proposed (without the deduplication of the Grgit call) didn't experience this failure; do you think we could merge the commitId improvement and follow up on the rat implicit dependencies in another PR if the latest commit doesn't fix the build?

**Comment by gharris1727** [comment]
It appears that the mustRunAfter directive has fixed the build, and now it's just normal flakey tests. I won't be reverting the fixups and should be ready for review.

**Comment by gharris1727** [comment]
@ijuma I'd still like to get this merged, please take another review pass, thanks!

**Comment by gharris1727** [comment]
The rat problem was fixed in #13854 so i'm removing my fix for it in this PR.

**Comment by gharris1727** [comment]
Hey @ijuma @divijvaidya @C0urante or @yashmayya Could you review this build improvement?

I'm still getting this `git gc` failure regularly when I try to build branches that have been idle for some time.
This also includes an improvement (the `isDirectory` guard) which allows us to use `git worktree` to check out and build multiple branches at one time.

I'd like to backport this to 3.4, 3.5, and 3.6.

**Review by ijuma — APPROVED** [review]
LGTM, thanks!

**Comment by ijuma** [comment]
Well, Jenkins failed - we need to firs that first.

**Comment by ijuma** [comment]
Build looks good, merging.

**Pull request merged**